contents/testing/writing-testable-code.md

```diff
@@ -0,0 +1,146 @@
+
+### Rule #2: Avoid Singleton Pattern
+
+Singleton Pattern, despite being a well-know design patter, are global states in sheep's clothing. They have **globally** accessible `getInstance()` method and a private singleton object (which is the global state).
+
+> Another problem about Singleton Pattern is that they are globally accessible, thus any method can access them inside the code without explicitly declaring in its API. In other words, the API lied about its dependency! - [Google Testing Blog](https://testing.googleblog.com/2008/08/by-miko-hevery-so-you-join-new-project.html)
+
+**Note** singletons (with lowercase s) are valid and sometimes very useful, which you can enforce their singleton identities by enforcing their constructors are called only once in the application. However, Singleton Pattern (with uppercase S) is almost always causing undesirable global states, which you should avoid. Only few examples of Singleton Pattern, including Constants and Logging, are acceptable since they don't affect the application logic.
+
+## Flaw #3: Dig Into Collaborators
+
+### Rule #1: Don't look for things. Ask for things.
+
+#### Warning Signs
+
+* Objects are passed in but never used directly (only used to get access to other objects)
+* [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) violation: method call chain walks an object graph with more than one dot. (e.g. `a.getB().getC().doSomething()`)
+
+The Law of Demeter can be summarized as "only talk to your **immediate** friends".
+
+Previously we mentioned the importance of dependency injection for creating seams in [Flaw #1: no seams for isolating the class under test](#flaw-1-no-seams-for-isolating-the-class-under-test). Here we emphasize on injecting only those **direct dependency**.
+
+Let's use an example to illustrate why **indirect** dependency hurts testability:
+
+Say you have an Authenticator class which needs a `Config` object, do you pass in a path to the configuration file, or just pass in a `Config` object? Since what you need is a `Config` object, not a file path, an example of indirect dependency will be you pass in a path to the configuration file, and read a `Config` object from the path. To test Authenticator, you have to first write some configuration to a file, pass the file path to Authenticator, and let it read a `Config` itself. Versus you can directly create a `Config` object and pass it to Authenticator using direct dependency.
+
+## Flaw #4: Class does too much
+
+To extreme, it becomes an anti-pattern, [God Object](https://en.wikipedia.org/wiki/God_object).
+
+### Rule #1: Avoid mixing of concerns
+
+#### Warning Signs:
+
+* summing up what the class does includes the word "and"
+* class has fields that are only used in some methods (objects hiding inside)
+* class has static methods that only operate on parameters (methods should be instance methods of one of the parameters)
```

Comment:
In some languages, it isn't possible to create methods that aren't associated with a class, e.g c# or java, so making them static is the only choice, even if they only use their parameters and don't touch other state.